fix: support no changes

gmeligio opened the pull request:
Currently, if a working directory has no changes, the step will finish with a failure. This is unexpected since nothing wrong happened, and it's just that there is no need to make a commit. That no-changes can be handled by not running the making the Github API request. In that case, there would be no `commit-reponse`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/github-api-commit-action?pullRequest=67) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/github-api-commit-action?pullRequest=67) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/github-api-commit-action?pullRequest=67) it.</sub>

abannachGrafana commented on action.yml:
```diff
@@ -119,7 +119,8 @@ runs:
         if [[ "$additions" == "" && "$deletions" == "" ]]; then
           echo "No changes to commit"
-          exit 1
+          echo "any_changed=false"  >> $GITHUB_OUTPUT
+          exit 0
```
Could we add an input to the action? Something like `succeed-if-no-changes` which defaults to `false` preserving the existing functionality. My thought is since the intent of the action is to commit changes. In general, if there are no files staged to be committed then something went wrong in the steps leading up to where this action would be called. 

I agree though that in some workflows it may be beneficial to allow this to pass without failing since that might be ok for the workflow at hand. That's where allowing an exit code of `0` if `succeed-if-no-changes` is `true` would be great.

gmeligio replied:
I'd think of the action like a `git push`, which doesn't fail without changes. It's true that changing this behavior would be a breaking change for the action.

Disregarding the breaking change, as a git user, I would expect an opposite input variable `fail-no-changes` if I want to validate that there were changes pushed.

abannachGrafana commented:
I'll need to investigate why it's failing to run the tests. This is happening on another PR in the repo so I don't believe it's localized to your branch.

abannachGrafana reviewed (changes requested):
Please add optional input of `succeed-if-no-changes` with a default of `false` to control the exit code. This will preserve existing functionality while supporting an option where no changes committed is ok in a workflow where that's ok.

gmeligio commented:
> @gmeligio I've merged #68 to hopefully resolve the test run issue. Please update your branch and we'll see if the tests run successfully.

Thank you @abannachGrafana. Rebased but the workflows need to be approved.

gmeligio commented:
> Please add optional input of `succeed-if-no-changes` with a default of `false` to control the exit code. This will preserve existing functionality while supporting an option where no changes committed is ok in a workflow where that's ok.

Thank you for reviewing @abannachGrafana . My idea comes from using the git CLI and the [stefanzweifel/git-auto-commit-action](https://github.com/stefanzweifel/git-auto-commit-action) action which don't fail if there are no changes.

`grafana/github-api-commit-action` it's better than what I knew before since it pushes verified commits and it's easier to configure.

Aside from my proposal being a breaking change, I'd think that detecting if there were changes in the working dir can be done with a separate step like `git status` or `git diff` or even with the [tj-actions/changed-files](https://github.com/tj-actions/changed-files) action. So we could even not create any variable at all.

I understand your suggestion of compatibility and we can still go with `success-if-no-changes` but I think that could be the default expected as a git user, and if any change detection is supported in the action, it might be the opposite `fail-if-no-changes`.

Before proceeding with the change, what do you think?

abannachGrafana commented:
Re: the workflow runs

Thanks! The other PR dev also did so and we still have a failure due to repo permissions. I'm in progress on a fix for that, but am waiting on a request to go through.

> I think that could be the default expected as a git user

Correct as a git user in the terminal you would realize the error and could correct it if there were no changes staged. Since github actions are an automated workflow with no way to intervene I still believe the action should fail by default if the implementor forgot or failed to change files when its expected that there should be a change. 

Consider something like an "Update changelog" workflow that is supposed to add PR info to the changelog. If the changelog wasn't changed or the workflow creator forgot to stage the file in a step. They may think everything is fine if the commit step was successful.

> a separate step like git status or git diff or even with the [tj-actions/changed-files](https://github.com/tj-actions/changed-files?rgh-link-date=2025-04-09T19%3A53%3A05Z) action

Yes! I'd pose that this should be done prior and to skip calling the commit step all together with a [conditional](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif) on the step. I'd also think that if a workflow may or may not make changes, it'd be better structured in that way as well so that it's clear it's possible that nothing is committed.

I'd like to stick with the `success-if-no-changes` flag. This preserves backwards compatibility and the intention of the action.

gmeligio commented:
Got it. Thank you for the explanation @abannachGrafana . I added `success-if-no-changes`. It's ready for another review.

The `test.yml` workflow is running now but it's not running the two tests I added.

abannachGrafana commented:
Yeah, I'm fighting with the tests because the generated GITHUB_TOKEN doesn't have write permission with event type of `pull_request`. I've been trying a bunch of stuff. So far no luck. Will need to give up for today, but I did merge a change so that it won't trigger tests on forked prs. It's just unfortunate.

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          ref: ${{ github.head_ref || github.ref }}
```
```suggestion
```

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+        continue-on-error: true
+        with:
+          token: ${{ github.token }}
+          stage-all-files: true
```
```suggestion
```

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+        run: |
+          git push --force --delete origin ${{ steps.setup-test-branch.outputs.branch-name }}
+      
+
```
```suggestion
```

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          ref: ${{ github.head_ref || github.ref }}
```
```suggestion
```

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+      - name: Setup test branch
+        id: setup-test-branch
+        run: |
+          BRANCH_NAME="test_no-changes-$(date +%s)"
```
```suggestion
          BRANCH_NAME="test_success_no-changes-$(date +%s)"
```

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+        continue-on-error: true
+        with:
+          token: ${{ github.token }}
+          stage-all-files: true
```
```suggestion
```

abannachGrafana commented on .github/workflows/test.yml:
```diff
@@ -429,4 +429,100 @@ jobs:
+      - name: Setup test branch
+        id: setup-test-branch
+        run: |
+          BRANCH_NAME="test_no-changes-$(date +%s)"
```
```suggestion
          BRANCH_NAME="test_failure_no-changes-$(date +%s)"
```

abannachGrafana reviewed (changes requested):
Looking great. The tests make sense too. Just a few suggestions

gmeligio commented:
Thanks. Applied the suggestions.

abannachGrafana commented:
thanks @gmeligio !

gmeligio commented:
Hey @abannachGrafana, could you please release this?

abannachGrafana commented:
@gmeligio New release `0.4.0` has been created. If you find yourself in a bind next time and need to reference changes to an action that were merged to main, but hasn't had a proper release yet you can either use `<action>@main` (or whatever the default branch is) or `<action>@<commit-hash>` and that should allow you to use the latest code.

gmeligio commented:
Thank you! I got it today with renovate. I'll take what you mentioned into account for next time.
